### PR TITLE
BoxDecoration tweaks

### PR DESCRIPTION
**Pull request by Hixie**
Reorder the code, remove some redundant documentation, improve the
interpolation logic.

### CODE REVIEW

File: packages/flutter/lib/src/painting/box_decoration.dart [hunk 1]
```diff
@@ -79,7 +79,7 @@ class BoxDecoration extends Decoration {
     this.boxShadow,
     this.gradient,
     this.shape: BoxShape.rectangle,
-  });
+  }) : assert(shape != null);
```
Comment (cbracken) [inline packages/flutter/lib/src/painting/box_decoration.dart]: Is it worth explicitly documenting that `[shape] must not be null`?

**Review by cbracken — APPROVED** [review]

**Comment by cbracken** [comment]
lgtm

File: packages/flutter/lib/src/painting/box_decoration.dart (hunk 1 above)
Comment (Hixie) [reply packages/flutter/lib/src/painting/box_decoration.dart]: Fixed. Also improved other documentation while I was at it.

**Comment by Hixie** [comment]
Will land on green.

**Comment by Hixie** [comment]
actually, landing on red, because redness seems to just be a flake.

**Pull request merged**